features/mogelijkOnjuist.feature

```diff
@@ -280,10 +281,27 @@ Functionaliteit: Mogelijk onjuist
         """
       En bevat het antwoord geen property "mogelijkOnjuist.huisnummer"
 
+Onderstaande scenario's zijn erop gericht dat afnemers altijd de noodzakelijke mogelijk onjuist informatie ontvangen als deze door toepassing van de fieldsparameter in principe niet zou worden geleverd.
+
+Scenario: leveren mogelijkOnjuist wanneer een woonplaats geometrie mogelijk onjuist is en vragen expand van woonplaats bij resource adressen
+      Gegeven bij object Woonplaats is “geometrie” in onderzoek
+      Als de resource adressen wordt opgevraagd met parameter fields=<property> en de gevraagde properties zijn niet woonplaats en/of woonplaatsIdentificatie en/of adresregel2
+      En met parameter expand=woonplaats of expand=woonplaats.<property>
+      Dan bevat het antwoord:
+        """
+            "mogelijkOnjuist": {
+                "woonplaatsIdentificatie": true,
```

Comment:
Even een check. Tegen welke OAS specificatie moet deze feature file worden aangehouden? Ik gebruik die van de master en daar zie ik dat WoonplaatsMogelijkOnjuist de property geometrie heeft. Ik verwacht dan dat deze op true wordt gezet en niet woonplaatsIdentificatie die daar niet is gedefinieerd

Reply:
Hier wordt /adressen endpoint bevraagd en wordt woonplaats ge-expand. Hier komt dus een AdresUitgebreid resource terug met daarin AdresMogelijkOnjuist en niet een WoonplaatMogelijkOnjuist.
De geometrie zit in de _embedded van woonplaats. Vanaf het /adressen endpoint kan woonplaats ge-expand worden maar niet de geometrie. Een mogelijk onjuiste woonplaats geometrie kan betekenen dat een adres in een andere woonplaats kan liggen. Daarom wordt de mogelijkOnjuist.woonplaatsIdentificatie = true teruggegeven.

Reply:
Als ik door de componenten navigeer (AdresHalCollectie > AdresHal > AdresEmbedded > WoonplaatsHalBasis > Woonplaats > WoonplaatsMogelijkOnjuist kom ik een geometrie property tegen. Regel 2690 in P.R. #424. Kijk ik niet naar hetzelfde?

Reply:
Ik ben eerlijk gezegd lost. Ik expand woonplaats. Woonplaats heeft een property van het type WoonplaatsMogelijkOnjuist met property geometrie. Geometrie is mogelijk onjuist. Waarom kan deze niet naar true en moet in AdresMogelijkOnjuist de woonplaatsIdentificatie property op true? De enige manier om te snappen wat er aan de hand is, is middels de toelichting. Ik denk niet dat een consumer hier logica op kan programmeren.

Reply:
Volgens het scenario op regel 200 wordt als de geometrie van de woonplaats niet wordt gevraagd er ook geen mogelijk onjuist van de woonplaats geleverd. In dit geval kun je de geometrie niet bevragen, want dat zou een dubbele expand zijn. De afnemer ontvangt dus een antwoord waarvan hij niet weet dat het mogelijk onjuist is. Dit scenario (en alles vanaf regel 284) is een soort van vangnetscenario om te voorkomen dat de afnemer gegevens gebruikt waarvan hij niet weet dat er misschien iets mee aan de hand is.

Reply:
Als ik het begrijp zeggen we dus het volgende:

Gegeven geometrie van woonplaats is in onderzoek

Als GET /adressen?...&fields=straat&expand=woonplaats
Dan
```
[
  {
    "straat": "...",
    "mogelijkOnjuist": {
       "woonplaatsIdentificatie": true
    ",
    "_embedded": {
      "woonplaats": {
        ...
      }
    }
  }
]
```

De scenario die ik nu beschrijf is een consumer/afnemer die **alleen** de woonplaats van adressen wil hebben die aan een criteria voldoen. Door een design flaw in de fields query parameter kan hij niet aangeven dat hij geen enkele property wil van een Adres, dus geeft hij een field op om niet te overfetchen.
In dit geval moet de consumer die alleen geïnteresseerd is in de woonplaats in code toch twee niveaus omhoog om te kunnen bepalen of er iets niet juist is aan de woonplaats. Dit voelt niet goed. Code technisch zeker niet, want de consumer wordt naar mijn mening gedwongen om code principes als Separation of Concern en Single Responsibility los te laten.

Wat ik mij afvraag is hoe de response eruit zou zien in de volgende situatie eruit zal zien:

Gegeven geometrie van woonplaats is in onderzoek
En status van woonplaats is in onderzoek
Als GET /adressen?...&fields=straat&expand=woonplaats.naam

Is de response dan:

```
[
  {
    "straat": "...",
    "mogelijkOnjuist": {
       "woonplaatsIdentificatie": true,
       "toelichting": [
         "geometrie is in onderzoek",
         "status is in onderzoek"
      ]
    ",
    "_embedded": {
      "woonplaats": {
        "naam": "..."
      }
    }
  }
]
```

of wordt de response:

```
[
  {
    "straat": "...",
    "mogelijkOnjuist": {
       "woonplaatsIdentificatie": true,
       "toelichting": [
         "geometrie is in onderzoek",
      ]
    ",
    "_embedded": {
      "woonplaats": {
        "naam": "...",
        "mogelijkOnjuist": {
           "status": true,
           "toelichting": [
              "status is in onderzoek"
           ]
        }
      }
    }
  }
]
```

Is de volgende response dan niet duidelijker en consistenter?

```
[
  {
    "straat": "...",
    "_embedded": {
      "woonplaats": {
        "naam": "...",
        "mogelijkOnjuist": {
           "geometrie": "true",
           "status": true,
           "toelichting": [
              "geometrie is in onderzoek",
              "status is in onderzoek"
           ]
        }
      }
    }
  }
]
```

Reply:
Voortbordurend waarom naar mijn mening dit niet correct is. Geometrie is gewoon een property van Woonplaats net als status. We hebben er voor gekozen om het te embedden om te voorkomen dat je voor elke woonplaats zijn geometrie ook meekrijgt als er geen fields wordt gebruikt. Wij hebben hier dus het embedden misbruikt, want geometrie is geen (sub)resource.

Wat mij brengt bij het volgend punt: een consumer kan via /adressen geen geometrie van een woonplaats ophalen. Dat betekent dat hij dat via /woonplaatsen endpoint doen. Echter /woonplaatsen ondersteunt geen bevraging met meerdere woonplaatsidentificaties. Dit betekent dus dat de consumer dan voor elke woonplaats die hij heeft opgehaald  via /adressen een request moet doen op /woonplaatsen als hij van de woonplaatsen de geometrie wil. Hetzelfde probleem als het ophalen van adresseerbaarobjecten met één pandIdentificatie.

Reply:
Eens Melvin. Ik vind de 3e respons duidelijker. Je zou de 2e krijgen. 
Deze constructie wordt veroorzaakt door het scenario op regel 199 welke volgens mij is bedoeld om niet teveel informatie te geven. Als we dat nou gewoon niet doen. Hoeven we voor woonplaatsen deze uitzonderingen/vangnetregeling bij expand woonplaats en _links niet toe te passen.

Reply:
Ik weet niet wat je precies bedoeld met : Hetzelfde probleem als het ophalen van adresseerbaarobjecten met één pandIdentificatie.

Reply:
> 
> 
> Ik weet niet wat je precies bedoeld met : Hetzelfde probleem als het ophalen van adresseerbaarobjecten met één pandIdentificatie.

We hebben /adresseerbareobjecten moeten uitbreiden met de pandIdentificaties query parameter omdat het kan voorkomen dat een consumer na het ophalen van panden de bijbehorende adresseerbare objecten wil ophalen. Zonder de pandIdentificaties parameter moet hij dat per pand doen in plaats van in een batch.

Hetzelfde kan ook van toepassing zijn voor woonplaatsen ophalen via adressen. Als een consumer van elke opgehaalde woonplaats de geometrie nodig heeft, dan moet hij, met de huidige /woonplaatsen endpoint, per woonplaats doen. Als we dit niet willen dan moeten we een nieuwe /woonplaatsen endpoint toevoegen met woonplaatsIdentificaties query parameter zodat dit in een batch kan.

Reply:
> Als een consumer van elke opgehaalde woonplaats de geometrie nodig heeft

Is daar een user story voor? We hebben destijds juist bewust ervoor gekozen geometrie niet standaard in de woonplaats resource op te nemen omdat die erg groot is. Is het dan wenselijk om in één keer van een lijst woonplaatsen de geometrie op te halen? Ik zie zelf nog niet direct waarom je vanuit het zoeken van adressen direct de geometrie van de woonplaatsen zou willen hebben. Hoe dan ook, dat is denk ik een ander onderwerp.

Ik denk dat je wel een goed punt hebt wat betreft Separation of Concern en Single Responsibility en dat je voorstel beter is. Dan moet er dus in de embedded resource worden aangegeven dat die embedded resource mogelijk niet de juiste is. Ik denk alleen dat dan elk geleverd gegeven van de embedded resource mogelijk onjuist is, en dat als toelichting de toelichting die hoort bij de relatie (ligt in/ligt aan/geometrie) wordt gegeven.

Bijvoorbeeld /adressen/1234207890123456?fields=straat&expand=woonplaats als geometrie in onderzoek is
```
[
  {
    "straat": "...",
    "_embedded": {
      "woonplaats": {
        "identificatie": "1234",
        "domein": "NL.IMBAG.Woonplaats",
        "naam": "voorbeeld"
        "status": "Woonplaats aangewezen ",
        "mogelijkOnjuist": {
           "naam": true
           "status": true
           "toelichting": [
              "Geometrie of woonplaatsgrens is mogelijk onjuist, waardoor gaten of overlappingen ontstaan in de registratie van woonplaatsen. Gevolg kan zijn dat een object in een verkeerde woonplaats, in twee woonplaatsen, of in geen enkele woonplaats ligt."
           ]
        }
      }
    }
  }
]
```

Bijvoorbeeld bij het vragen van specifiek veld van de woonplaats /adressen/1234207890123456?fields=straat&expand=woonplaats.status als geometrie in onderzoek is
```
[
  {
    "straat": "...",
    "_embedded": {
      "woonplaats": {
        "status": "Woonplaats aangewezen ",
        "mogelijkOnjuist": {
           "status": true
           "toelichting": [
              "Geometrie of woonplaatsgrens is mogelijk onjuist, waardoor gaten of overlappingen ontstaan in de registratie van woonplaatsen. Gevolg kan zijn dat een object in een verkeerde woonplaats, in twee woonplaatsen, of in geen enkele woonplaats ligt."
           ]
        }
      }
    }
  }
]
```

Bijvoorbeeld /adressen/1234207890123456?fields=straat&expand=woonplaats.status als openbare ruimte ligt in gerelateerde woonplaats in onderzoek is
```
[
  {
    "straat": "...",
    "_embedded": {
      "woonplaats": {
        "status": "Woonplaats aangewezen ",
        "mogelijkOnjuist": {
           "status": true
           "toelichting": [
              "Mogelijk verkeerde woonplaats gebruikt. De straat moet verwijzen naar de woonplaats waarin de straat fysiek ligt."
           ]
        }
      }
    }
  }
]
```

Hoe gaan we hiermee dan om wanneer alleen een gegeven wordt gevraagd waar (nu) geen mogelijkOnjuist veld gevraagd is? Leveren we dan alleen mogelijkOnjuist toelichting? Bijvoorbeeld bij het vragen van specifiek veld van de woonplaats /adressen/1234207890123456?fields=straat&expand=woonplaats.documentnummer als geometrie in onderzoek is:
```
[
  {
    "straat": "...",
    "_embedded": {
      "woonplaats": {
        "documentnummer": "2012-123456",
        "mogelijkOnjuist": {
           "toelichting": [
              "Geometrie of woonplaatsgrens is mogelijk onjuist, waardoor gaten of overlappingen ontstaan in de registratie van woonplaatsen. Gevolg kan zijn dat een object in een verkeerde woonplaats, in twee woonplaatsen, of in geen enkele woonplaats ligt."
           ]
        }
      }
    }
  }
]
```

Reply:
> 
> 
> > Als een consumer van elke opgehaalde woonplaats de geometrie nodig heeft
> 
> Is daar een user story voor? We hebben destijds juist bewust ervoor gekozen geometrie niet standaard in de woonplaats resource op te nemen omdat die erg groot is.

Klopt. Geef ik ook aan in mijn rant.

> Is het dan wenselijk om in één keer van een lijst woonplaatsen de geometrie op te halen? Ik zie zelf nog niet direct waarom je vanuit het zoeken van adressen direct de geometrie van de woonplaatsen zou willen hebben. Hoe dan ook, dat is denk ik een ander onderwerp.
>

Op dit moment niet. Het is puur een observatie van mij. Ik zie mogelijk hetzelfde optreden als bij adres waar de adresseerbare objecten gewenst zijn en bij pand waar ook behoefte is aan adresseerbare objecten. Gelukkig is het beperkt tot slechts één gegeven. Anders is de kans veel groter dat dit gewenste functionaliteit wordt.

> Ik denk dat je wel een goed punt hebt wat betreft Separation of Concern en Single Responsibility en dat je voorstel beter is. Dan moet er dus in de embedded resource worden aangegeven dat die embedded resource mogelijk niet de juiste is. Ik denk alleen dat dan elk geleverd gegeven van de embedded resource mogelijk onjuist is, en dat als toelichting de toelichting die hoort bij de relatie (ligt in/ligt aan/geometrie) wordt gegeven.
> 
> Bijvoorbeeld /adressen/1234207890123456?fields=straat&expand=woonplaats als geometrie in onderzoek is
> 
> ```
> [
>   {
>     "straat": "...",
>     "_embedded": {
>       "woonplaats": {
>         "identificatie": "1234",
>         "domein": "NL.IMBAG.Woonplaats",
>         "naam": "voorbeeld"
>         "status": "Woonplaats aangewezen ",
>         "mogelijkOnjuist": {
>            "naam": true
>            "status": true
>            "toelichting": [
>               "Geometrie of woonplaatsgrens is mogelijk onjuist, waardoor gaten of overlappingen ontstaan in de registratie van woonplaatsen. Gevolg kan zijn dat een object in een verkeerde woonplaats, in twee woonplaatsen, of in geen enkele woonplaats ligt."
>            ]
>         }
>       }
>     }
>   }
> ]
> ```
> 
> Bijvoorbeeld bij het vragen van specifiek veld van de woonplaats /adressen/1234207890123456?fields=straat&expand=woonplaats.status als geometrie in onderzoek is
> 
> ```
> [
>   {
>     "straat": "...",
>     "_embedded": {
>       "woonplaats": {
>         "status": "Woonplaats aangewezen ",
>         "mogelijkOnjuist": {
>            "status": true
>            "toelichting": [
>               "Geometrie of woonplaatsgrens is mogelijk onjuist, waardoor gaten of overlappingen ontstaan in de registratie van woonplaatsen. Gevolg kan zijn dat een object in een verkeerde woonplaats, in twee woonplaatsen, of in geen enkele woonplaats ligt."
>            ]
>         }
>       }
>     }
>   }
> ]
> ```
> 
> Bijvoorbeeld /adressen/1234207890123456?fields=straat&expand=woonplaats.status als openbare ruimte ligt in gerelateerde woonplaats in onderzoek is
> 
> ```
> [
>   {
>     "straat": "...",
>     "_embedded": {
>       "woonplaats": {
>         "status": "Woonplaats aangewezen ",
>         "mogelijkOnjuist": {
>            "status": true
>            "toelichting": [
>               "Mogelijk verkeerde woonplaats gebruikt. De straat moet verwijzen naar de woonplaats waarin de straat fysiek ligt."
>            ]
>         }
>       }
>     }
>   }
> ]
> ```
> 
> Hoe gaan we hiermee dan om wanneer alleen een gegeven wordt gevraagd waar (nu) geen mogelijkOnjuist veld gevraagd is? Leveren we dan alleen mogelijkOnjuist toelichting? Bijvoorbeeld bij het vragen van specifiek veld van de woonplaats /adressen/1234207890123456?fields=straat&expand=woonplaats.documentnummer als geometrie in onderzoek is:
> 
> ```
> [
>   {
>     "straat": "...",
>     "_embedded": {
>       "woonplaats": {
>         "documentnummer": "2012-123456",
>         "mogelijkOnjuist": {
>            "toelichting": [
>               "Geometrie of woonplaatsgrens is mogelijk onjuist, waardoor gaten of overlappingen ontstaan in de registratie van woonplaatsen. Gevolg kan zijn dat een object in een verkeerde woonplaats, in twee woonplaatsen, of in geen enkele woonplaats ligt."
>            ]
>         }
>       }
>     }
>   }
> ]
> ```

Ik heb eerlijk gezegd nooit begrepen waarom een consumer tot op property niveau moet/wil weten dat een gegeven mogelijkOnjuist is of niet. Als ik de toelichting bij geometrie is onjuist lees, dan zou ik niet alleen de geometrie maar eigenlijk de gehele woonplaats binnen de context van het adres als onjuist moeten beschouwen. Ik vraag mij dus echt af of wij dingen niet onnodig complex hebben gemaakt. Mijn gevoel zegt mij dat 80% van de consumers het al voldoende vinden als ze weten dat de woonplaats mogelijk onjuist is en niet welk property van woonplaats.

Reply:
Ik vind het niet onlogisch dat je hoort welke property er echt onjuist is. In de verschillende usecases is verschillende informatie wel of niet relevant. Met de extra informatie kan de medewerker een afweging maken of hij het gegeven wel of niet gebruikt of eerst verder onderzoek doet. 

Als ik terugga naar waar de discussie mee startte is dat de vraag waarom op die plek woonplaatsidentificatie wordt teruggegeven. We hebben geconstateerd dat we naar AdresMogelijkOnjuist kijken. In deze opbouw klopt dat. 

Dan is de vraag of we deze opbouw willen. 
Bij alle vangnetscenario's maakt de property geen onderdeel uit van het object dat met expand of _links wordt opgevraagd. Daar kan de informatie dus niet worden getoond. De mogelijk onjuist wordt daarom gegeven vanuit de resource die wordt gebruikt voor het request en die de mogelijk onjuist informatie bevat.

Bij woonplaatsgeometrie maakt de property juist wel onderdeel uit van het met expand of _links gevraagde object, maar leveren we de informatie niet omdat we standaard geen geometrie leveren. Om fouten bij de afnemer te voorkomen moet je dus:
1. de informatie leveren op de plek die nu als voorstel in de featurefile staat, of,
2.  bij de woonplaats. Als je dat doet moet je het scenario vanaf regel 200 schrappen en het altijd leveren ook als je geen woonplaats geometrie levert.

Reply:
OK. optie 2 is dan naar mijn mening de enige correcte oplossing, dus bij de woonplaats.